File: Components/WinForms/CS/SchedulerMainDemo/Forms/MyInplaceEditor.cs
using System;
using System.Windows.Forms;

namespace DevExpress.XtraScheduler.Demos.Forms {
    public partial class MyInplaceEditor : DevExpress.XtraEditors.XtraForm {
        Appointment appointment;
        SchedulerControl control;

        public MyInplaceEditor() {
            InitializeComponent();
            SubscribeKeyDownEvents();
        }
        public event EventHandler CommitChanges;
        public event EventHandler RollbackChanges;

        private void SubscribeKeyDownEvents() {
            appointmentLabelEdit.KeyDown += new KeyEventHandler(AppointmentLabelEdit_KeyDown);
            edtDescription.KeyDown += new KeyEventHandler(Editor_KeyDown);
            edtSubject.KeyDown += new KeyEventHandler(Editor_KeyDown);
        }
        public void AppointmentLabelEdit_KeyDown(object sender, KeyEventArgs e) {
            if (!appointmentLabelEdit.IsPopupOpen)
                Editor_KeyDown(sender, e);
        }
        void Editor_KeyDown(object sender, KeyEventArgs e) {
            switch(e.KeyCode) {
                case Keys.Enter:
                    e.Handled = true;
                    OnCommitChanges();
                    break;
                case Keys.Escape:
                    e.Handled = true;
                    OnRollbackChanges();
                    break;
            }
        }

        void OnCommitChanges() {
            if(CommitChanges != null)
                CommitChanges(this, EventArgs.Empty);
        }

        void OnRollbackChanges() {
            if(RollbackChanges != null)
                RollbackChanges(this, EventArgs.Empty);
        }
        protected override void OnShown(EventArgs e) {
            ISchedulerStorageBase storage = control.DataStorage;
            if(storage.Appointments.IsNewAppointment(appointment)) {
                edtSubject.SelectionLength = 0;
                edtSubject.SelectionStart = edtSubject.Text.Length;
            }
            base.OnShown(e);
        }
      
        public void FillForm(SchedulerControl control, Appointment appointment) {
            this.appointment = appointment;
            this.control = control;
            ISchedulerStorage storage = control.DataStorage;
            this.appointmentLabelEdit.Storage = storage;
            this.appointmentLabelEdit.AppointmentLabel = storage.Appointments.Labels.GetById(appointment.LabelKey);
            this.edtSubject.Text = appointment.Subject;
            this.edtDescription.Text = appointment.Description;
        }
        public void ApplyChanges() {
            appointment.Subject = edtSubject.Text;
            appointment.Description = edtDescription.Text;
            appointment.LabelKey = control.DataStorage.Appointments.Labels.IndexOf(appointmentLabelEdit.AppointmentLabel);            
        }
    }
}

File: Components/WinForms/VB/SchedulerMainDemo/Forms/MyInplaceEditor.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows.Forms

Namespace DevExpress.XtraScheduler.Demos.Forms
	Partial Public Class MyInplaceEditor
		Inherits DevExpress.XtraEditors.XtraForm
		Private appointment As Appointment
		Private control As SchedulerControl

		Public Sub New()
			InitializeComponent()
			SubscribeKeyDownEvents()
		End Sub
		Public Event CommitChanges As EventHandler
		Public Event RollbackChanges As EventHandler

		Private Sub SubscribeKeyDownEvents()
			AddHandler appointmentLabelEdit.KeyDown, AddressOf AppointmentLabelEdit_KeyDown
			AddHandler edtDescription.KeyDown, AddressOf Editor_KeyDown
			AddHandler edtSubject.KeyDown, AddressOf Editor_KeyDown
		End Sub
		Public Sub AppointmentLabelEdit_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
			If (Not appointmentLabelEdit.IsPopupOpen) Then
				Editor_KeyDown(sender, e)
			End If
		End Sub
		Private Sub Editor_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
			Select Case e.KeyCode
				Case Keys.Enter
					e.Handled = True
					OnCommitChanges()
				Case Keys.Escape
					e.Handled = True
					OnRollbackChanges()
			End Select
		End Sub

		Private Sub OnCommitChanges()
			RaiseEvent CommitChanges(Me, EventArgs.Empty)
		End Sub

		Private Sub OnRollbackChanges()
			RaiseEvent RollbackChanges(Me, EventArgs.Empty)
		End Sub
		Protected Overrides Sub OnShown(ByVal e As EventArgs)
			Dim storage As ISchedulerStorageBase = control.DataStorage
			If storage.Appointments.IsNewAppointment(appointment) Then
				edtSubject.SelectionLength = 0
				edtSubject.SelectionStart = edtSubject.Text.Length
			End If
			MyBase.OnShown(e)
		End Sub

		Public Sub FillForm(ByVal control As SchedulerControl, ByVal appointment As Appointment)
			Me.appointment = appointment
			Me.control = control
			Dim storage As ISchedulerStorage = control.DataStorage
			Me.appointmentLabelEdit.Storage = storage
			Me.appointmentLabelEdit.AppointmentLabel = storage.Appointments.Labels.GetById(appointment.LabelKey)
			Me.edtSubject.Text = appointment.Subject
			Me.edtDescription.Text = appointment.Description
		End Sub
		Public Sub ApplyChanges()
			appointment.Subject = edtSubject.Text
			appointment.Description = edtDescription.Text
			appointment.LabelKey = control.DataStorage.Appointments.Labels.IndexOf(appointmentLabelEdit.AppointmentLabel)
		End Sub
	End Class
End Namespace

